Support for subjects in SQLiteDB

sbassi opened the pull request:
Here is the support for the subjects in the DB. There is also the updated "starter DB".

eseiver commented on allofplos/makedb.py:
```diff
@@ -20,11 +21,10 @@
-# diving into the code more thoroughly 
-corpusdir = 'allofplos/allofplos_xml'
+# diving into the code more thoroughly
+corpusdir = os.path.join(os.path.dirname(__file__), 'allofplos_xml')
```
The corpus dir can be anywhere on the file system, so it should be `corpusdir = get_corpusdir()`

eseiver commented on allofplos/makedb.py:
```diff
@@ -105,19 +108,24 @@ class PLOSArticle(BaseModel):
+                          'starter_corpus','*.xml'))
 else:
-    allfiles = os.listdir(corpusdir)
+    allfiles = glob.glob(os.path.join(corpusdir, '*.xml'))
```
Instead of `glob`, you should find that the `Corpus` class will serve this purpose, and you can pass in the directory if it's not the corpus directory. Specifically, `Corpus().iter_files()` is a generator will let you iterate through every article without needing to generate the list of files beforehand.

sbassi replied:
I am using glob because I want to get only xml files, and this directory has a __init__.py file I would grab if I request all of them. Will check iter_files() method.

mpacer commented on allofplos/makedb.py:
```diff
@@ -155,6 +163,25 @@ class CoAuthorPLOSArticle(BaseModel):
+    taxonomy_set = set()
+    taxonomy = article.taxonomy
+    if taxonomy:
+        for values in taxonomy.values():
```
If taxonomy is guaranteed to be a dict, you can remove remove the conditional.

sbassi replied:
OK, I was thinking about an empty result, I think correction articles has no taxonomy, not sure right now, will check by running this on the whole corpus.

eseiver commented on allofplos/makedb.py:
```diff
@@ -7,6 +7,7 @@
 
 import argparse
 import datetime
+import glob
```
This can be excluded now

eseiver commented on allofplos/makedb.py:
```diff
@@ -17,14 +18,15 @@
 
+from .corpus import Corpus
 from .transformations import filename_to_doi, convert_country
+from . import get_corpus_dir
```
you can import starterdir from . too

eseiver replied:
I actually don't think you need to import `get_corpus_dir` or set a `corpusdir` value because the Corpus class already calls it

eseiver commented on allofplos/makedb.py:
```diff
@@ -105,26 +110,33 @@ class PLOSArticle(BaseModel):
 
 if args.starter:
-    allfiles = os.listdir('starter_corpus')
+    starter_dir = os.path.join(os.path.dirname(__file__), 'starter_corpus')
```
import `starterdir` from allofplos' __init__ instead

eseiver commented on allofplos/makedb.py:
```diff
@@ -105,26 +110,33 @@ class PLOSArticle(BaseModel):
+    allfiles = corpus.iter_files
 else:
-    allfiles = os.listdir(corpusdir)
+    corpus = Corpus(corpusdir)
```
defaults to `get_corpus_dir()` so you can just use `corpus = Corpus()`. only need to specify for `starterdir`

sbassi replied:
ok, I called with get_corpus_dir because I read the init method of Corpus class and was None.
On a second inspection (after this comment), I realized that you are changing the corpusdir value when there is no value, so this is the same of declaring a default value, but with the problem that is not reflected in the method signature (you need the read the code to see it, when could be clear by just looking at the signature). So I am changing the init with this.
Looks at the changed init and tell me if I am missing something.

mpacer commented on allofplos/corpus/corpus_class.py:
```diff
@@ -9,10 +9,8 @@
     """A collection of PLOS articles."""
 
-    def __init__(self, directory=None, extension='.xml'):
+    def __init__(self, directory=get_corpus_dir(), extension='.xml'):
```
This is not a pattern that I support having, __anywhere__ in the codebase. The set the default to `None` and then modify it if it is `None` is a classic pythonic  for addressing this kind of case. 

If you want a description of why we shouldn't do this, check out: 
the "Default arguments evaluation" section of  http://blog.thedigitalcatonline.com/blog/2015/02/11/default-arguments-in-python/.

mpacer replied:
> default arguments are evaluated at definition time (for example when you first import the module)

meaning that it's no longer going to be appropriately flexible for people who don't want to use a environmental variable directly.

sbassi replied:
I see it now, so I will revert this change.

mpacer commented on allofplos/makedb.py:
```diff
@@ -105,26 +104,32 @@ class PLOSArticle(BaseModel):
+    allfiles = corpus.iter_files
 else:
-    allfiles = os.listdir(corpusdir)
+    corpus = Corpus()
```
This should still work even if you don't pass in `get_corpus_dir()` because it is the default in `Corpus`, and that should work using `None` as the default value.

mpacer replied:
If it doesn't work, please explain… because that means there's a bug elsewhere.

sbassi replied:
It is working, I am not passing get_corpus_dir() anymore.

mpacer commented on allofplos/makedb.py:
```diff
@@ -105,26 +104,32 @@ class PLOSArticle(BaseModel):
 if args.starter:
-    allfiles = os.listdir('starter_corpus')
+    corpus = Corpus(starterdir)
+    allfiles = corpus.iter_files
```
I think the naming here is a little confusing since it follows a different usage pattern in the three cases. 

is `args.starter` `None` if it is not included?  if so, couldn't you do

```python
corpus_dir = starterdir if args.starter else None
allfiles = Corpus(corpus_dir).iter_files
files = iter(random.sample(list(corpus.iter_files, args.random))) if args.random else allfiles

for file_ in tqdm(files):
⋮
```

with the possibility of changing the second ternary to a full conditional block if its too long for your taste. 

but that would make it more concise and would allow combining a random sample with starterdir in one step.

mpacer replied:
I'll just make a PR against your branch to change this

sbassi replied:
it is OK for me, personally I find easy to read the named and nested option, but I can handle any version, so please go ahead.

mpacer replied:
It's that it's not allfiles when you're sampling… and while I was thinking about it I realised that the Corpus creation didn't need to be its own step since you weren't using the Corpus for anything else.

sbassi commented:
yes, the name is misleading.

eseiver commented:
LGTM, feel free to merge after any last changes you need @mpacer

mpacer commented:
Could you do an interactive rebase to cleanup your fixup commits? 

I just did one locally and if you did it on the commits you currently have here, your rebase instructions would read:

```
 pick df8b942 Support for subjects in SQLiteDB
 f dc95888 addressing PR suggestions
 pick 2825744 New starter DB with subjects
 f daa339e Adding suggestions from PR
 f 2a93e66 removed unused name
 f 8dd7ea0 removed unused name and comment
 f 3df0ecd Reverting chances
```

mpacer commented:
If you'd prefer that I do the rebase, just ask and I'll push the changes, I just don't want to overwrite your branch without your explicit permission.

sbassi commented:
I did the rebase (git rebase -i upstream/master), is there anything else I should do?

mpacer commented:
you need to force push onto your branch to overwrite the previous commits

mpacer commented:
or more accurately … `git push --force-with-lease origin`

sbassi commented:
I got this:

git push --force-with-lease origin
warning: push.default is unset; its implicit value has changed in
Git 2.0 from 'matching' to 'simple'. To squelch this message
and maintain the traditional behavior, use:

  git config --global push.default matching

To squelch this message and adopt the new behavior now, use:

  git config --global push.default simple

When push.default is set to 'matching', git will push local branches
to the remote branches that already exist with the same name.

Since Git 2.0, Git defaults to the more conservative 'simple'
behavior, which only pushes the current branch to the corresponding
remote branch that 'git pull' uses to update the current branch.

See 'git help config' and search for 'push.default' for further information.
(the 'simple' mode was introduced in Git 1.7.11. Use the similar mode
'current' instead of 'simple' if you sometimes use older versions of Git)

Everything up-to-date

mpacer commented:
Git 2.8 (released in 2016) removed this warning… are you able to update your version of git? The future is lovely if you can! 

If you can't… then you could just set that setting… unless elsewhere you rely on the old matching behaviour?

But what's weird is that it comes up at all since you specified a remote…

if you do `git push --force-with-lease origin subjectsDB` does that work?

mpacer commented:
Also… I'm assuming that `origin` is `sbassi` for you and `upstream` is `PLOS`… is that the case?

sbassi commented:
I am on 2.7.4, when I do Everything up-to-date I get:
Everything up-to-date

mpacer commented:
2.7.4 < 2.8 so it makes sense that you'd be getting that warning. 

What did you try to do? Everything up-to-date isn't a git command.

sbassi commented:
this is the answer when I try the force push you showed me. I am on the
move now, will try again tomorrow.

On Feb 27, 2018 5:32 PM, "M Pacer" <notifications@github.com> wrote:

> 2.7.4 < 2.8 so it makes sense that you'd be getting that warning.
>
> What did you try to do? Everything up-to-date isn't a git command.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PLOS/allofplos/pull/83#issuecomment-369089450>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOjtdg64X9EYdpwgLfDM1MdF7znK2Gaks5tZKzDgaJpZM4SJH7a>
> .
>

mpacer commented:
@sbassi can you run `git remote -v`?

sbassi commented:
$ git remote -v
origin	git@github.com:sbassi/allofplos.git (fetch)
origin	git@github.com:sbassi/allofplos.git (push)
upstream	https://github.com/PLOS/allofplos.git (fetch)
upstream	https://github.com/PLOS/allofplos.git (push)

mpacer commented:
So `git push --force-with-lease origin subjectsDB` gives you `Everything up-to-date`?

Can you run `git reflog show|head` and paste here what you've got?

sbassi commented:
Yes to the first question:

~/projects/plos/sbrepo/allofplos$ git push --force-with-lease origin subjectsDB
Everything up-to-date

Will run the other command now

sbassi commented:
:~/projects/plos/sbrepo/allofplos$ git reflog show|head
3df0ecd HEAD@{0}: rebase -i (finish): returning to refs/heads/subjectsDB
3df0ecd HEAD@{1}: rebase -i (start): checkout upstream/master
3df0ecd HEAD@{2}: commit: Reverting chances
8dd7ea0 HEAD@{3}: commit: removed unused name and comment
2a93e66 HEAD@{4}: commit: removed unused name
daa339e HEAD@{5}: commit: Adding suggestions from PR
2825744 HEAD@{6}: commit: New starter DB with subjects
dc95888 HEAD@{7}: checkout: moving from getcorpusdirdoc to subjectsDB
e908d9b HEAD@{8}: checkout: moving from master to getcorpusdirdoc
e908d9b HEAD@{9}: checkout: moving from subjectsDB to master

sbassi commented:
I the idea is to flatten the commit history, I can create a new branch in another directory and copy the changed files in that branch, so the net result will be the current changes only.

mpacer commented:
Ok it looks like your rebase didn't work. 

When you ran the `git rebase -i upstream/master` command, a text editor should have opened up and you should have changed the file that was created to look like the examples I gave you above. E.g., 

```
 pick df8b942 Support for subjects in SQLiteDB
 f dc95888 addressing PR suggestions
 pick 2825744 New starter DB with subjects
 f daa339e Adding suggestions from PR
 f 2a93e66 removed unused name
 f 8dd7ea0 removed unused name and comment
 f 3df0ecd Reverting chances
```

> I the idea is to flatten the commit history, I can create a new branch in another directory and copy the changed files in that branch, so the net result will be the current changes only.

I'd prefer to work through how to do this as it will make it easier for you to do this in the future.

Also, after you rebase, I'd appreciate if you would merge my PR on your repo… https://github.com/sbassi/allofplos/pull/1 unless you object to it in which case I'd ask you to review it and raise those objections.

sbassi commented:
Here is my rebase:

~/projects/plos/sbrepo/allofplos$ git rebase -i upstream/master
Successfully rebased and updated refs/heads/subjectsDB.

mpacer commented:
Ah… `git fetch upstream` and then do that again.

mpacer commented:
something is severely wrong with that rebase…do you want me to just do the rebase and overwrite what you have here?

Debugging this remotely is going to be waaaaay more complicated that I had realised

sbassi commented:
yes please

mpacer commented:
Ok, the rebase worked. do you want to look at my PR on your branch now?

mpacer commented:
Thanks a tonne! Let's figure out a time to talk through an interactive rebase because I think once you've done it once it'll seem a lot easier. 

We should look into making a full subpackage for database stuff since there are now a few files that are tightly connected that are living in the top level of the package.
